ulib/libax/src/cbindings/socket.rs

```diff
@@ -0,0 +1,337 @@
+        match self {
+            Socket::Udp(udpsocket) => Ok(udpsocket.send(buf)?),
+            Socket::Tcp(tcpsocket) => Ok(tcpsocket.send(buf)?),
+        }
+    }
+
+    pub fn recv(&self, buf: &mut [u8]) -> LinuxResult<usize> {
+        match self {
+            Socket::Udp(udpsocket) => Ok(udpsocket.recv_from(buf).map(|e| e.0)?),
+            Socket::Tcp(tcpsocket) => Ok(tcpsocket.recv(buf)?),
+        }
+    }
+
+    fn bind(&mut self, addr: SocketAddr) -> LinuxResult {
+        match self {
+            Socket::Udp(udpsocket) => Ok(udpsocket.bind(addr)?),
+            Socket::Tcp(tcpsocket) => Ok(tcpsocket.bind(addr)?),
+        }
+    }
+
+    fn connect(&mut self, addr: SocketAddr) -> LinuxResult {
+        match self {
+            Socket::Udp(udpsocket) => Ok(udpsocket.connect(addr)?),
+            Socket::Tcp(tcpsocket) => Ok(tcpsocket.connect(addr)?),
+        }
+    }
+
+    fn sendto(&self, buf: &[u8], addr: SocketAddr) -> LinuxResult<usize> {
+        match self {
+            // diff: must bind before sendto
+            Socket::Udp(udpsocket) => Ok(udpsocket.send_to(buf, addr)?),
+            Socket::Tcp(_) => Err(LinuxError::EISCONN),
+        }
+    }
+
+    fn recvfrom(&self, buf: &mut [u8]) -> LinuxResult<(usize, Option<SocketAddr>)> {
+        match self {
+            // diff: must bind before recvfrom
+            Socket::Udp(udpsocket) => {
+                Ok(udpsocket.recv_from(buf).map(|res| (res.0, Some(res.1)))?)
+            }
+            Socket::Tcp(tcpsocket) => Ok(tcpsocket.recv(buf).map(|res| (res, None))?),
+        }
+    }
+
+    fn listen(&mut self) -> LinuxResult {
+        match self {
+            Socket::Udp(_) => Err(LinuxError::EOPNOTSUPP),
+            Socket::Tcp(tcpsocket) => Ok(tcpsocket.listen()?),
+        }
+    }
+
+    fn accept(&mut self) -> LinuxResult<TcpSocket> {
+        match self {
+            Socket::Udp(_) => Err(LinuxError::EOPNOTSUPP),
+            Socket::Tcp(tcpsocket) => Ok(tcpsocket.accept()?),
+        }
+    }
+
+    fn shutdown(&self) -> LinuxResult {
+        match self {
+            Socket::Udp(udpsocket) => {
+                udpsocket.peer_addr()?;
+                udpsocket.shutdown().unwrap();
+                Ok(())
+            }
+
+            Socket::Tcp(tcpsocket) => {
+                tcpsocket.peer_addr()?;
+                tcpsocket.shutdown().unwrap();
+                Ok(())
+            }
+        }
+    }
+}
+
+pub(super) fn stat_socket(_socket: &Socket) -> AxResult<ctypes::stat> {
+    // not really implemented
+    let st_mode = ((0o14_u32) << 12) | (0o777_u32);
+    Ok(ctypes::stat {
+        st_ino: 1,
+        st_nlink: 1,
+        st_mode,
+        st_uid: 0,
+        st_gid: 0,
+        st_size: 0,
+        st_blocks: 0,
+        st_blksize: 512,
+        ..Default::default()
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ax_socket(domain: c_int, socktype: c_int, protocol: c_int) -> c_int {
```

Comment:
Print debug info about arguments for all `ax_*` functions, refers to file.rs. Same below.